src/main/java/ch/njol/skript/sections/EffSecShoot.java

```diff
@@ -0,0 +1,201 @@
+	protected @Nullable TriggerItem walk(Event event) {
+		lastSpawned = null;
+		Number finalVelocity = velocity != null ? velocity.getSingle(event) : DEFAULT_SPEED;
+		Direction finalDirection = direction != null ? direction.getSingle(event) : Direction.IDENTITY;
+		if (finalVelocity == null || finalDirection == null)
+			return null;
+
+		for (Object shooter : shooters.getArray(event)) {
+			for (EntityData<?> entityData : types.getArray(event)) {
+				Entity finalProjectile = null;
+				Vector vector;
+				if (shooter instanceof LivingEntity livingShooter) {
+					vector = finalDirection.getDirection(livingShooter.getLocation()).multiply(finalVelocity.doubleValue());
+					Class<? extends Entity> type = entityData.getType();
+					if (Fireball.class.isAssignableFrom(type)) {
+						if (trigger != null) {
+							livingShooter.getWorld().spawn(
+								livingShooter.getEyeLocation().add(vector.clone().normalize().multiply(0.5)),
+								type,
+								(Consumer) getSpawnConsumer(event, vector, entityData, livingShooter)
+							);
+						} else {
+							Fireball fireball = (Fireball) livingShooter.getWorld().spawn(
+								livingShooter.getEyeLocation().add(vector.clone().normalize().multiply(0.5)),
+								type
+							);
+							fireball.setShooter(livingShooter);
+							finalProjectile = fireball;
+						}
+					} else if (Projectile.class.isAssignableFrom(type)) {
+						if (trigger != null) {
+							livingShooter.launchProjectile(
+								(Class<? extends Projectile>) type,
+								vector,
+								(Consumer) getSpawnConsumer(event, vector, entityData, livingShooter)
+							);
+						} else {
+							finalProjectile = livingShooter.launchProjectile((Class<? extends Projectile>) type);
+							set(finalProjectile, entityData);
+						}
+					} else {
+						Location location = livingShooter.getLocation();
+						location.setY(location.getY() + livingShooter.getEyeHeight() / 2);
+						if (trigger != null) {
+							entityData.spawn(location, (Consumer) getSpawnConsumer(event, vector, entityData, livingShooter));
+						} else {
+							finalProjectile = entityData.spawn(location);
+						}
+					}
+				} else {
+					vector = finalDirection.getDirection((Location) shooter).multiply(finalVelocity.doubleValue());
+					if (trigger != null) {
+						entityData.spawn((Location) shooter, (Consumer) getSpawnConsumer(event, vector, entityData, null));
+					} else {
+						finalProjectile = entityData.spawn((Location) shooter);
+					}
+				}
+				if (finalProjectile != null) {
+					finalProjectile.setVelocity(vector);
+					lastSpawned = finalProjectile;
+				}
+			}
+		}
+
+		return super.walk(event, false);
+	}
+
+	@SuppressWarnings("unchecked")
+	private static <E extends Entity> void set(Entity entity, EntityData<E> entityData) {
+		entityData.set((E) entity);
+	}
+
+	private Consumer<? extends Entity> getSpawnConsumer(Event event, Vector vector, EntityData<?> entityData, @Nullable LivingEntity shooter) {
```

Comment:
i feel the method name should actually include what the consumer does, not that it's a consumer cuz that's visible in the return type. something like `afterSpawn`